[Remote Inspection] Add some more SPI to reset visibility adjustment and count adjustment rects

whsieh opened the pull request:
#### 1b90f5339ab042ec882acaed926f8bd0acc2d74c
<pre>
[Remote Inspection] Add some more SPI to reset visibility adjustment and count adjustment rects
<a href="https://bugs.webkit.org/show_bug.cgi?id=271901">https://bugs.webkit.org/show_bug.cgi?id=271901</a>

Reviewed by Aditya Keerthi.

Add support for a couple of new methods on `WKWebView`. See below for more detail.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setVisibilityAdjustment):
(WebCore::Element::addVisibilityAdjustment): Deleted.

Rename `addVisibilityAdjustment` -&gt; `setVisibilityAdjustment`, and make it replace the rare data&apos;s
visibility adjustment flags (instead of only adding to them). This is necessary, since we now need
a way to reset this flag and restore the original element state.

* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::setVisibilityAdjustment):
(WebCore::ElementRareData::addVisibilityAdjustment): Deleted.

More renaming — see above.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::inflatedClientRectForAdjustmentRegionTracking):

Pull this out into a static helper, so that we can also use it in `resetVisibilityAdjustments` to
update the current visibility adjustment regions, to leave out elements that are no longer adjusted.

(WebCore::adjustVisibilityIfNeeded):

Adopt `setVisibilityAdjustment`.

(WebCore::ElementTargetingController::adjustVisibility):
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):

Keep track of all the elements in the top document that have been adjusted, with a new member
`m_adjustedElements`. This is updated when the client explicitly triggers visibility adjustment
through API, and also when we adjust visibility automatically for repeatedly targeted regions.

(WebCore::ElementTargetingController::resetAdjustmentRegions):

Also reset `m_adjustedElements`.

(WebCore::ElementTargetingController::resetVisibilityAdjustments):

Use `m_adjustedElements` to undo any visibility adjustments for elements that correspond to the list
of element identifiers (or, if the list is empty, all current adjusted elements).

(WebCore::ElementTargetingController::numberOfVisibilityAdjustmentRects const):

Use `m_adjustedElements` to estimate the number of visually distinct adjustment rects, relative to
client coordinates. Note that this intentionally treats fully overlapped element regions as the same
rect, for the purposes of reporting the final count.

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(elementsFromWKElements):
(-[WKWebView _resetVisibilityAdjustmentsForTargetedElements:completionHandler:]):
(-[WKWebView _adjustVisibilityForTargetedElements:completionHandler:]):
(-[WKWebView _numberOfVisibilityAdjustmentRectsWithCompletionHandler:]):

Add plumbing for these new methods.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::extractIdentifiers):
(WebKit::WebPageProxy::resetVisibilityAdjustmentsForTargetedElements):
(WebKit::WebPageProxy::adjustVisibilityForTargetedElements):
(WebKit::WebPageProxy::numberOfVisibilityAdjustmentRects):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::resetVisibilityAdjustmentsForTargetedElements):
(WebKit::WebPage::numberOfVisibilityAdjustmentRects):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(-[WKWebView numberOfVisibilityAdjustmentRects]):
(-[WKWebView resetVisibilityAdjustmentsForTargets:]):
(TestWebKitAPI::TEST(ElementTargeting, NearbyOutOfFlowElements)):

Augment an existing API test to cover these two new methods.

Canonical link: <a href="https://commits.webkit.org/276850@main">https://commits.webkit.org/276850@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42e891fdb5ef743d8fd04603f5736f8d53fdeb2d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41930 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48198 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29315 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22418 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37551 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46470 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18728 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19492 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40683 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3936 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42189 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41034 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50334 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20886 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17379 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44685 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43565 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10184 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/22547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21879 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -648,6 +656,8 @@ void ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions(Doc
     if (!bodyElement)
         return;
 
+    document.updateLayoutIgnorePendingStylesheets();
```
Do we need to be concerned about this triggering script execution? Or is that no longer the case since `LocalFrameView::performPostLayoutTasks` has `ScriptDisallowedScope` now?

whsieh replied:
I'm _pretty sure_ it's safe to update layout underneath this codepath, since other rendering update steps can trigger layout and script execution as well (and `Page::updateRendering` itself triggers layout). That said, I can omit this part from the patch for now (since it's not strictly required to implement the new methods), and add it back in a followup if needed.

I suspect there's a way to trigger a bug in this code by invalidating layout in a prior rendering update step (e.g. under `runResizeSteps`, `runScrollSteps` or `flushAutofocusCandidates`) and cause this logic to ignore potential targets.

whsieh committed:
[Remote Inspection] Add some more SPI to reset visibility adjustment and count adjustment rects
https://bugs.webkit.org/show_bug.cgi?id=271901

Reviewed by Aditya Keerthi.

Add support for a couple of new methods on `WKWebView`. See below for more detail.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setVisibilityAdjustment):
(WebCore::Element::addVisibilityAdjustment): Deleted.

Rename `addVisibilityAdjustment` -> `setVisibilityAdjustment`, and make it replace the rare data's
visibility adjustment flags (instead of only adding to them). This is necessary, since we now need
a way to reset this flag and restore the original element state.

* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::setVisibilityAdjustment):
(WebCore::ElementRareData::addVisibilityAdjustment): Deleted.

More renaming — see above.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::inflatedClientRectForAdjustmentRegionTracking):

Pull this out into a static helper, so that we can also use it in `resetVisibilityAdjustments` to
update the current visibility adjustment regions, to leave out elements that are no longer adjusted.

(WebCore::adjustVisibilityIfNeeded):

Adopt `setVisibilityAdjustment`.

(WebCore::ElementTargetingController::adjustVisibility):
(WebCore::ElementTargetingController::adjustVisibilityInRepeatedlyTargetedRegions):

Keep track of all the elements in the top document that have been adjusted, with a new member
`m_adjustedElements`. This is updated when the client explicitly triggers visibility adjustment
through API, and also when we adjust visibility automatically for repeatedly targeted regions.

(WebCore::ElementTargetingController::resetAdjustmentRegions):

Also reset `m_adjustedElements`.

(WebCore::ElementTargetingController::resetVisibilityAdjustments):

Use `m_adjustedElements` to undo any visibility adjustments for elements that correspond to the list
of element identifiers (or, if the list is empty, all current adjusted elements).

(WebCore::ElementTargetingController::numberOfVisibilityAdjustmentRects const):

Use `m_adjustedElements` to estimate the number of visually distinct adjustment rects, relative to
client coordinates. Note that this intentionally treats fully overlapped element regions as the same
rect, for the purposes of reporting the final count.

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(elementsFromWKElements):
(-[WKWebView _resetVisibilityAdjustmentsForTargetedElements:completionHandler:]):
(-[WKWebView _adjustVisibilityForTargetedElements:completionHandler:]):
(-[WKWebView _numberOfVisibilityAdjustmentRectsWithCompletionHandler:]):

Add plumbing for these new methods.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::extractIdentifiers):
(WebKit::WebPageProxy::resetVisibilityAdjustmentsForTargetedElements):
(WebKit::WebPageProxy::adjustVisibilityForTargetedElements):
(WebKit::WebPageProxy::numberOfVisibilityAdjustmentRects):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::resetVisibilityAdjustmentsForTargetedElements):
(WebKit::WebPage::numberOfVisibilityAdjustmentRects):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(-[WKWebView numberOfVisibilityAdjustmentRects]):
(-[WKWebView resetVisibilityAdjustmentsForTargets:]):
(TestWebKitAPI::TEST(ElementTargeting, NearbyOutOfFlowElements)):

Augment an existing API test to cover these two new methods.

Canonical link: https://commits.webkit.org/276850@main
